### DISCUSSION TITLE
Save / Load form into HasOne relationship

### DISCUSSION THREAD

**Discussion by shiroamada** [filamentphp/filament#6496] [Help] [answered]
I can't find any documentation how to do so.

I have 2 Models `User` and `UserInfo`. The relationship is 1 on 1.

User.php
```
    public function user_info()
    {
        return $this->hasOne(UserInfo::class);
    }
```

UserInfo.php
```
    public function user()
    {
        return $this->belongsTo(User::class);
    }
```
In the tableform, I am able to retrieve the data
```
 public static function table(Table $table): Table
    {
        return $table
            ->columns([
                Tables\Columns\TextColumn::make("name"),
                Tables\Columns\TextColumn::make("email"),
                Tables\Columns\TextColumn::make("user_info.mobile"),
                Tables\Columns\TextColumn::make("user_info.status"),
            ])
```

But in the form, I can't save it and load the data.
```
  public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\TextInput::make("name")->required(),
                Forms\Components\TextInput::make("email")
                    ->email()
                    ->required(),
                Forms\Components\TextInput::make("user_info.mobile")
                    ->numeric()
                    ->required(),
```

I search for new post.
https://github.com/filamentphp/filament/pull/2304
but no idea how to archive it.
If I add
```
                Forms\Components\TextInput::make("user_info.mobile")
                     ->relationship("user_info", "user_id")
                    ->numeric()
                    ->required(),
                Forms\Components\Select::make("user_info.status")
                    ->options([
                        "active" => "Active",
                        "inactive" => "Inactive",
                    ])
                    ->relationship("user_info", "user_id")
                    ->required(),
```
I got php error
```
Filament\Forms\Components\Select::getRelationship(): Return value must be of type Illuminate\Database\Eloquent\Relations\BelongsTo|Illuminate\Database\Eloquent\Relations\BelongsToMany|Znck\Eloquent\Relations\BelongsToThrough|null, Illuminate\Database\Eloquent\Relations\HasOne returned
```
<img width="1344" alt="image" src="https://github.com/filamentphp/filament/assets/2062946/5dc77cbf-b197-46d9-8b40-0a8fd284b999">

**Comment by RX-77IT** [comment, marked as answer]
What about using a Fieldset?
```php
Forms\Components\Fieldset::make('User Info')
    ->relationship('user_info')
    ->columns(1)
    ->schema([
        Forms\Components\TextInput::make("mobile")
            ->numeric()
            ->required(),
        Forms\Components\Select::make("status")
            ->required()
            ->options([
                "active" => "Active",
                "inactive" => "Inactive",
            ]),
    ]),
```
Reference: https://filamentphp.com/docs/2.x/admin/resources/getting-started#hasone

**Reply by danharrin** [reply]
Or any other layout component works too with `relationship()`

**Reply by aldriguz** [reply]
Sorry if this is the wrong place to ask this question, but I have an error with `BelontsTo/HasOne` relationship using Fieldset, this repo shows up the same issue that I have (https://github.com/Uolsen/filament-hasone-bug). 

When using like this, is not populating the `user_id` column of the parent (`ClientResource`)

```php
Forms\Components\Fieldset::make('Usuario')
                ->relationship('user')
                ->schema([
                    Forms\Components\TextInput::make('name')->label('Usuario')->required(),
                    Forms\Components\TextInput::make('email')->label('Email')->email()->required(),
                    Forms\Components\TextInput::make('password')->label('Contraseña')->password()->required(),
                ]),
```

On submit, missing 'user_id' on the query, I'm expecting to store the childs first and then use the new ids to populate in parent query.

<img width="1202" alt="image" src="https://github.com/filamentphp/filament/assets/3867738/49517a4d-30e0-4070-95b6-924d91224fc0">

**Reply by etgohomeok** [reply]
@aldriguz Did you ever figure out a solution to this? I'm having the same issue with a Fieldset relationship that is creating the child record but not updating the id in the parent record.

**Reply by danharrin** [reply]
Make it nullable and we will save it after its been created

**Reply by etgohomeok** [reply]
@danharrin Thank you for responding!

In my instance, the field is nullable. The form creates the records without any error, but the key in the parent record is null (ie. it does not get updated to point at the id of the child record.

For what it's worth, if I change the relationship from HasOne to BelongsTo then everything works normally - perhaps this is just me misunderstanding the correct kind of relationship to use in this scenario,

Nonetheless I'll include some details below. For reference, "spec" is my parent record and "typeset_spec" is the child.

spec migration:

```php
$table->foreignId('typeset_spec_id')->nullable()->constrained('jc_typeset_specs');
```

Spec.php:

```php
public function typesetSpec() {
    return $this->hasOne(TypesetSpec::class, 'id', 'typeset_spec_id');
}
```

SpecResource.php:

```php
Fieldset::make('typesetting')
    ->relationship('typesetSpec')
    ->label('Typesetting')
    ->schema([
        Forms\Components\TextInput::make('font')
            ->label('Font'),
        Forms\Components\TextInput::make('font_size')
            ->label('Font Size')
            ->numeric(),
    ]),
```

The form (stripped down example):

![image](https://github.com/filamentphp/filament/assets/19737434/fdfa7575-bcdb-4120-9ef1-eaa3dcb00cf5)

Parent table result:

![image](https://github.com/filamentphp/filament/assets/19737434/bca2209f-c16e-4094-a9d2-ba41fbeeea98)

Child table result:

![image](https://github.com/filamentphp/filament/assets/19737434/a38bfedc-7290-476b-8b69-bd0df8c02329)

**Reply by danharrin** [reply]
Is the key `$fillable`?

**Reply by etgohomeok** [reply]
For both models I am using `$guarded` so I believe the `typeset_spec_id` field should be fillable.

Spec.php and TypesetSpec.php both have:

```php
protected $guarded = [
    'id',
    'created_at',
    'updated_at',
];
```

**Reply by danharrin** [reply]
Please create a new project, reproduce the error, and create an issue with the reproduction attached

**Reply by etgohomeok** [reply]
Hi, just popping in to give some closure to my issue - I was in the process of putting together a reproduction repo when I figured out that the problem was with me confusing how HasOne and BelongsTo relationships are meant to be structured in Laravel.

In HasOne the foreign key should be in the child model (ie. *not* the model that contains the relationship), while in BelongsTo the foreign key is in the model that contains the relationship. I had this mixed up (which was why it suddenly started working when I changed the relationship to BelongsTo without changing my database structure).

If anyone else is having a similar issue, I encourage you to do some research on the difference between HasOne and BelongsTo relationships and make sure your database migrations and models are structured correctly - if they're not following proper convention the relationship may still "just work" everywhere until you get to the point where Filament is having issues creating things.
